### ISSUE TITLE
metaseq-train hangs at global barrier when 2-nodes launch.

### ISSUE THREAD

**Issue by ftgreat** [facebookresearch/metaseq#652]
### Before asking:   
1. search the issues.   
2. search the docs.    

#### What is your question?
I run metaseq-train command to start training on 2-nodes & 1 gpu per node. 
But hangs at global barrier when first loading json dataset. The same script runs ok on single node & 2 gpus.
![image](https://user-images.githubusercontent.com/10432020/219582179-55d02e68-fd3a-4290-a5bf-6b95d89020eb.png)

#### Code
metaseq-train --task streaming_language_modeling \
  data-bin/pile-00 \
  --vocab-filename data-bin/pile-00/gpt2-vocab.json \
  --merges-filename data-bin/pile-00/gpt2-merges.txt \
  --criterion cross_entropy \
  --batch-size 4 \
  --save-dir checkpoints/lm_transformer_pile-00 \
  --arch transformer_lm --share-decoder-input-output-embed \
  --dropout 0.1 \
  --optimizer adam --weight-decay 0.01 --clip-norm 0.0 \
  --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 1 --warmup-init-lr 1e-07 \
  --tokens-per-sample 256 --sample-break-mode none \
  --fp16 \
  --bf16 \
  --dataset-impl lazy \
  --num-workers 4 \
  --decoder-learned-pos \
  --log-format json \
  --max-epoch 1 \
  --arch transformer_lm \
  --distributed-world-size 2 \
  --distributed-init-method tcp://$MASTER_ADDR:$MASTER_PORT \
  --distributed-rank $RANK

   

#### What have you tried?
The same script runs ok on single node & 2 gpus.
I added detailed logs and found torch.distributed.is_initialized is True and the global barrier hangs. 

#### What's your environment?

 - metaseq Version (e.g., 1.0 or master):
 - PyTorch Version (e.g., 1.0)
 - OS (e.g., Linux):
 - How you installed metaseq (`pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by ftgreat** [comment]
@suchenzang So sorry to bother you. Could you please give me some adivce about this? Thanks.

**Comment by suchenzang** [comment]
Could you turn on your NCCL info logs and paste everything you see in a gist? See: https://github.com/pytorch/pytorch/issues/23074 for an example

**Comment by ftgreat** [comment]
> 
After turning on NCCL info, got these errors "Got completion from peer 172.24.6.167<52542> with error 12, opcode 0, len 0, vendor err 129".

Master Logs:
2023-02-20 02:15:13 | WARNING | metaseq.distributed.utils | Distributed is_initialized True
2023-02-20 02:15:13 | WARNING | metaseq.distributed.utils | Distributed get_rank 0
2023-02-20 02:15:13 | WARNING | metaseq.distributed.utils | call global_barrier, torch.distributed.is_initialized True
2023-02-20 02:15:13 | WARNING | metaseq.distributed.utils | call global_barrier, rank 0

job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-master-0:3038:3839 [0] transport/net_ib.cc:881 NCCL WARN NET/IB : Got completion from peer 172.24.72.162<52752> with error 12, opcode 0, len 0, vendor err 129
job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-master-0:3038:3839 [0] NCCL INFO include/net.h:28 -> 2
job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-master-0:3038:3839 [0] NCCL INFO transport/net.cc:459 -> 2
job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-master-0:3038:3839 [0] NCCL INFO proxy.cc:351 -> 2
job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-master-0:3038:3839 [0] NCCL INFO proxy.cc:452 -> 2 [Proxy Thread]


Worker Logs:
2023-02-20 02:15:13 | WARNING | metaseq.distributed.utils | Distributed is_initialized True
2023-02-20 02:15:13 | WARNING | metaseq.distributed.utils | Distributed get_rank 1
2023-02-20 02:15:13 | WARNING | metaseq.distributed.utils | call global_barrier, torch.distributed.is_initialized True
2023-02-20 02:15:13 | WARNING | metaseq.distributed.utils | call global_barrier, rank 1

job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-worker-0:3007:3545 [0] transport/net_ib.cc:881 NCCL WARN NET/IB : Got completion from peer 172.24.6.167<52542> with error 12, opcode 0, len 0, vendor err 129
job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-worker-0:3007:3545 [0] NCCL INFO include/net.h:28 -> 2
job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-worker-0:3007:3545 [0] NCCL INFO transport/net.cc:459 -> 2
job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-worker-0:3007:3545 [0] NCCL INFO proxy.cc:351 -> 2
job-9e55b20a-baf5-4612-a9fb-a7b3c3effdd7-worker-0:3007:3545 [0] NCCL INFO proxy.cc:452 -> 2 [Proxy Thread]

**Comment by ftgreat** [comment]
@suchenzang Follow this issue https://github.com/karpathy/nanoGPT/pull/55 and try adding NCCL_IB_DISABLE=1, training process can start, but very slowly. When setting NCCL_IB_DISABLE=0, training process hangs still with same error. 

In our cluster, many training jobs run normally with NCCL_IB_DISABLE=0. Could you please give some advices about this.  Thanks

**Comment by ftgreat** [comment]
> @suchenzang Follow this issue [karpathy/nanoGPT#55](https://github.com/karpathy/nanoGPT/pull/55) and try adding NCCL_IB_DISABLE=1, training process can start, but very slowly. When setting NCCL_IB_DISABLE=0, training process hangs still with same error.
> 
> In our cluster, many training jobs run normally with NCCL_IB_DISABLE=0. Could you please give some advices about this. Thanks

After setting NCCL_IB_HCA, training is ok.
